asyncpg/pool.py

```diff
@@ -213,52 +212,71 @@ def __init__(self, pool, *, connect_args, connect_kwargs,
                     # an IO error, so terminate the connection.
                     self._con.terminate()
                 finally:
-                    self._con = None
                     raise ex
 
+        self._release()
+
         assert self._inactive_callback is None
         if self._max_inactive_time and self._con is not None:
             self._inactive_callback = self._pool._loop.call_later(
                 self._max_inactive_time, self._deactivate_connection)
 
-    async def close(self):
-        self._maybe_cancel_inactive_callback()
-        if self._con is None:
-            return
-        if self._con.is_closed():
-            self._con = None
+    async def wait_until_released(self):
+        if self._in_use is None:
             return
+        else:
+            await self._in_use
 
-        try:
+    async def close(self):
+        if self._con is not None:
+            # Connection.close() will call _release_on_close() to
+            # finish holder cleanup.
             await self._con.close()
-        finally:
-            self._con = None
 
     def terminate(self):
-        self._maybe_cancel_inactive_callback()
-        if self._con is None:
-            return
-        if self._con.is_closed():
-            self._con = None
-            return
-
-        try:
+        if self._con is not None:
+            # Connection.terminate() will call _release_on_close() to
+            # finish holder cleanup.
             self._con.terminate()
-        finally:
-            self._con = None
 
     def _maybe_cancel_inactive_callback(self):
         if self._inactive_callback is not None:
             self._inactive_callback.cancel()
             self._inactive_callback = None
 
     def _deactivate_connection(self):
-        assert not self._in_use
-        if self._con is None or self._con.is_closed():
-            return
-        self._con.terminate()
+        assert self._in_use is None
+        if self._con is not None:
+            self._con.terminate()
```

Comment:
Add a comment that we terminate the connection instead of just closing it because it's not in use.